Resend task statuses when reestablishing session

dperny opened the pull request:
Oh boy this one is a doozy

1. Makes the status reporter support batch updates
2. Makes the agent support batch updates as a status reporter
3. Makes the places that use a status report use batch updates when
available
4. Makes the worker's `Listen` method callable multiple times with the
same status reporter, reporting all of the statuses to that reporter
every time it is called.
5. Makes the agent call the worker's `Listen` method every time it
reestablishes a session.

This fixes a race condition that results from the Dispatcher in the
manager batching updates, and telling the agent that task updates have
succeeded when they haven't actually been committed to the raft log.
This has the effect of making the managers drop status updates if there
is a leadership change while a batch of status updates is in the
pipeline. The fix here is that when agent's reestablish a session, they
re-send all of their task statuses, so the manager will DEFINITELY be
up to date.

This commit is marked "WIP" because I'm not totally confident on the
architecture and I want some review before I write unit tests for all
this new junk.

/cc @stevvooe for review, and also @aluzzardi, because this is... 
quite an impactful change, and i'm not totally confident about this 
approach.

anshulpundir commented:
>This fixes a race condition that results from the Dispatcher in the
manager batching updates, and telling the agent that task updates have
succeeded when they haven't actually been committed to the raft log.

I want to understand this better. Why does the dispatcher tell the agent that the updates have succeeded when they haven't ? Is it possible to just either commit those updates to raft before ack-ing them, or to not ack at all ?

dperny commented:
@anshulpundir I added a comment to the dispatcher UpdateTaskStatus method explaining the issue.

anshulpundir commented on agent/agent.go:
```diff
@@ -285,6 +285,9 @@ func (a *Agent) run(ctx context.Context) {
 				}
+				// then, call Listen on the worker again to update all of the
+				// task statuses.
+				a.worker.Listen(ctx, reporter)
```
I suggest adding a function which gets all the updates instead of calling Listen() again, which is confusing to read.

dperny replied:
Yeah, will do this.

anshulpundir commented on agent/agent.go:
```diff
@@ -507,6 +511,63 @@ func (a *Agent) UpdateTaskStatus(ctx context.Context, taskID string, status *api
+	}
+}
+
+// UpdateTaskStatuses performs a bulk update of ALL task statuses provided.
```
nit: clarify this: performs a bulk update to the dispatcher.

anshulpundir commented on agent/agent.go:
```diff
@@ -507,6 +511,63 @@ func (a *Agent) UpdateTaskStatus(ctx context.Context, taskID string, status *api
+
+			var err error
+			for len(updates) > 0 {
+				// session.sendTaskStatuses can only send 1024 updates at a time.
```
Sending more than what the dispatcher can process seems suboptimal. Since this is a closed system, it's probably OK to make the agent aware of this limit and split the batch up on the sending side.

dperny replied:
I'd say "no" only because this is the pattern hinted at by the `session.sendTaskStatuses` method, and lets us ignore how large the batch CAN be in favor of just sending as much as we want and looping.

anshulpundir commented on agent/agent.go:
```diff
@@ -507,6 +511,63 @@ func (a *Agent) UpdateTaskStatus(ctx context.Context, taskID string, status *api
+				updates, err = session.sendTaskStatuses(ctx, updates...)
+				if err != nil {
+					// in the a.UpdateTaskStatus(), we skip the task unknown
+					// error, because we kno wwe shouldn't get it. we can't do
```
typos: kno wwe

anshulpundir commented on agent/reporter.go:
```diff
@@ -102,14 +122,43 @@ func (sr *statusReporter) run(ctx context.Context) {
+			// if the update was successful, then leave the map alone. all of
+			// the tasks are up to-date. go back to the top again so we can
+			// batch if we've had a bunch of tasks in the meantime
+			continue
```
I'd suggest putting an else instead of continue, somewhat easier to read.

anshulpundir commented on agent/worker.go:
```diff
@@ -34,7 +34,9 @@ type Worker interface {
 	// Listen to updates about tasks controlled by the worker. When first
 	// called, the reporter will receive all updates for all tasks controlled
-	// by the worker.
+	// by the worker. Listen can be called multiple times with the same
```
>Listen can be called multiple times
This is unrelated and should probably be addressed in a separate PR.

stevvooe commented on agent/agent.go:
```diff
@@ -483,7 +486,8 @@ func (a *Agent) withSession(ctx context.Context, fn func(session *session) error
 //
 // If an error is returned, the operation should be retried.
 func (a *Agent) UpdateTaskStatus(ctx context.Context, taskID string, status *api.TaskStatus) error {
```
Can we change this method to coordinate the batching rather than propagating the interface update through?

dperny replied:
The reason not to do so is because we end up having to locally cache batch members in more and more places, which is the pattern that has bit us and caused this bug in the first place. 

We'd end up with 1 task statuses map in the StatusReporter, and then another task status map to keep track of the batch in the Agent, and we have to synchronize access to both in their respective places, and it's just asking for trouble.

The interface change seems more impactful than it actually is because once I changed the interface, there was no reason not to go through and add the batch method up and down the stack. Like, once I added `UpdateTaskStatuses` on agent, i also added` UpdateTaskStatuses` on the status reporter because I already had it and there was the opportunity for a good optimization there.

stevvooe replied:
The maps have different functionality. One map is a per task register of the latest task. The other is a queue of statuses that need to be batched out. The rest of the code changes less if we just make this handle batching.

dperny commented:
@anshulpundir the biggest reason not to have the manager confirm that a task update is committed is because I'm substantially less confident that I can write that pattern correctly on the first try.

anshulpundir commented on manager/dispatcher/dispatcher.go:
```diff
@@ -584,6 +584,24 @@ func (d *Dispatcher) UpdateTaskStatus(ctx context.Context, r *api.UpdateTaskStat
 		validTaskUpdates = append(validTaskUpdates, u)
 	}
 
+	// NOTE(dperny): So, there is an issue here, a bad design decision. We add
```
nit: This is a great summary, but maybe adjust it  so it feels more like "how it works" rather than "what you did".

dperny replied:
I've reworded this slightly from when you made this comment, but not changed the specific line, which is why this looks unaddressed.

dperny commented:
@anshulpundir I'll update all of the tests when we've settled on the designed, and I'll remove WIP at that time.

anshulpundir commented:
I'd assume we've settled on this approach. I don't see any comments which indicate otherwise @dperny

dperny commented:
I've updated the PR. I removed the whole concept of a "Batchable" status reporter, and change the `statusReporterFunc`, which is a test helper type, to support batch updates. In the process, I caught a nasty bug where later task updates would get overwritten by earlier task updates if the batch send failed. I think that this change should cover unit testing for the reporter, but someone can feel free to weigh in with unit tests they'd like to see. I'm removing "WIP" from this PR, it's ready for real, final review.

stevvooe commented on agent/agent.go:
```diff
@@ -285,6 +285,9 @@ func (a *Agent) run(ctx context.Context) {
 				}
+				// then, call Listen on the worker again to update all of the
+				// task statuses.
+				a.worker.Report(ctx, reporter)
```
So, we trigger this on a full, incoming set? This will block the agent loop. Can we start a goroutine here?

nishanttotla commented on agent/agent.go:
```diff
@@ -285,6 +285,9 @@ func (a *Agent) run(ctx context.Context) {
 				if err := a.worker.Assign(ctx, msg.Changes); err != nil {
 					log.G(ctx).WithError(err).Error("failed to synchronize worker assignments")
 				}
+				// then, call Listen on the worker again to update all of the
```
which previous comment does this follow from?

anshulpundir commented on agent/reporter.go:
```diff
@@ -21,6 +22,18 @@ func (fn statusReporterFunc) UpdateTaskStatus(ctx context.Context, taskID string
 	return fn(ctx, taskID, status)
 }
 
+func (fn statusReporterFunc) UpdateTaskStatuses(ctx context.Context, statuses map[string]*api.TaskStatus) error {
```
Why is fn statusReporterFunc not an argument ?

dperny replied:
Because `fn` is the method receiver (it is an argument, just a special kind of argument).

stevvooe replied:
Allows a function to implement an interface.

anshulpundir commented:
Still reviewing test code. Should have review out soon.

dperny commented on manager/dispatcher/dispatcher.go:
```diff
@@ -629,6 +648,11 @@ func (d *Dispatcher) processUpdates(ctx context.Context) {
 		"method": "(*Dispatcher).processUpdates",
 	})
 
+	// TODO(dperny) we assume that if the batch fails, it means that the
```
Figure out if the batch send can fail without failing the dispatcher.

anshulpundir commented on agent/reporter_test.go:
```diff
@@ -21,17 +21,25 @@ func TestReporter(t *testing.T) {
-		mu       sync.Mutex
-		expected = make(map[string]*api.TaskStatus)
-		wg       sync.WaitGroup
+		// NOTE(dperny): don't increase this timeout! this test should never
```
Is there a way to test the batching code ? Successful batch send, splitting of a batch into multiple batches, failing and retrying the failed batch ?

dperny replied:
this tests all of those things from the reporter side by using an `UpdateTaskStatuses` method defined on the `statusReporterFunc`. `statusReporterFunc` fails sometimes at random, which will cause the reporter to retry. This test, with no changes, already caught 1 bug where in a failed batch, older tasks were rewriting newer ones.

cyli replied:
Non-blocking:  would you mind adding that as a comment to the test function, or as a comment on the random function?  (e.g. it non-deterministically tests the retry logic?)

cyli replied:
(alternately, would it make sense to make it fail every other call to `statusReporterFunc`, so we can guarantee that the retry functionality is tested?)

cyli commented on agent/reporter_test.go:
```diff
@@ -21,17 +21,25 @@ func TestReporter(t *testing.T) {
+		unique      = make(map[uniqueStatus]struct{})  // ensure we don't receive any status twice
+		mu          sync.Mutex
+		expected    = make(map[string]*api.TaskStatus)
+		// this test used to use a waitgroup, but i've removed it because if
```
Non-blocking nitpick:  this is an awesome comment to reviewers, but I'm not sure it is a useful comment in the code to say what the code used to do.  Perhaps just a comment at the top of the test saying that the failure condition is a deadlock?

stevvooe commented on agent/reporter.go:
```diff
@@ -13,6 +13,7 @@ import (
 // concurrently, so implementations should be goroutine-safe.
 type StatusReporter interface {
 	UpdateTaskStatus(ctx context.Context, taskID string, status *api.TaskStatus) error
+	UpdateTaskStatuses(ctx context.Context, statuses map[string]*api.TaskStatus) error
```
I am not sure this is a good change: the callers of this interface aren't typically supposed to deal with batching. I would prefer we leave this to a single task status and batch at the session level.

dperny replied:
Caller doesn't deal with batching. Caller provides either one task status or many task statuses. How they get processed is up to the implementation.

The best reason to not change the function definition to use the map everywhere is that I would have to go through every place that uses the function and change it, which makes backporting hard, testing hard, and code confidence hard.

stevvooe replied:
Ok, so let's put a TODO here. This should indicate that `UpdateTaskStatuses` is temporary and will be removed in favor of batching in `(*Agent).UpdateTaskStatus`.

The main problem with `UpdateTaskStatuses` is that it creates a lot of allocations that aren't necessary for the common cases, so we shouldn't move all calls to it.

dperny commented on agent/reporter.go:
```diff
@@ -102,27 +142,24 @@ func (sr *statusReporter) run(ctx context.Context) {
+		// current ones
+		sr.mu.Unlock()
+		err := sr.reporter.UpdateTaskStatuses(ctx, statuses)
+		sr.mu.Lock()
```
We don't need another check. At worst, we end up adding all of the updates back to the map, but when we next go through the loop, we'll see that the reporter is closed and exit.

dperny commented:
@anshulpundir PTAL

anshulpundir commented:
Also ping @nishanttotla @cyli @stevvooe for approval since they had previous comments.

cyli commented on agent/reporter.go:
```diff
@@ -102,27 +144,25 @@ func (sr *statusReporter) run(ctx context.Context) {
+		sr.mu.Unlock()
+		err := sr.reporter.UpdateTaskStatuses(ctx, statuses)
+		// re-lock the map so we can add statuses back if we need to.
+		sr.mu.Lock()
```
Non-blocking, just confirming my understanding: although it might be possible for the status reporter to be closed while `UpdateTaskStatuses` is executing, we don't care because adding statuses to a closed reporter won't cause any issues?

cyli commented on agent/reporter.go:
```diff
@@ -45,24 +58,52 @@ func newStatusReporter(ctx context.Context, upstream StatusReporter) *statusRepo
+// UpdateTaskStatuses updates all of the task statuses at once. The key
+// difference between this and calling UpdateTaskStatus repeatedly is that this
+// hold the lock until all updates are queued, meaning no update will happen
+// until all udpates are ready
```
very minor nit:  "updates" (vs udpates)

dperny commented:
@cyli addressed nits by fixing the spelling in one comment, and by adding a comment explaining why we don't need a second `sr.closed` check after reacquiring the lock in run.

dperny commented:
@stevvooe i'd rather not change `UpdateTaskStatus` to coordinate batching because it feels more clever to me, and while i might be clever enough to write it, i don't think i'm clever enough to debug it. i'd rather merge this as-is instead of letting perfect be the enemy of good enough.

anshulpundir commented:
> it seems like changing the UpdateTaskStatus function, if I'm understanding the conversation, which I may not be, can also be handled as add a new method signature, drop the old one, which would give us some time to convert the rest of the code over, too.

+1

dperny commented:
I'm poking at what @stevvooe wants and it's honestly not the best idea, or I'm otherwise misunderstanding it.

Batching solely at the session or agent level involves keeping a state in the session through which we accrue updates and then write them out after a critical period or count, whichever comes first. It introduces complexity by maintaining yet another local state.

Batching solely at the reporter level (for example, by having 1 method which takes only a map, or having 1 method which takes only a task id and status) has the trade off of either requiring a separate method call with the associated overhead for every task status update in an obvious batch, or by requiring a map allocation (or related optimization) for every update to with just 1 status. This is less bad than keeping a batch state in the agent or session, but still sub-optimal.

The current solution has two methods, one for updating a single status and one for sending more than one status at once. These two separate methods are only ever actually used on the `Reporter` object, which keeps local state of status updates, and then writes out batches at at time to the agent. The current solution, having two methods, lets the caller choose the most appropriate method for sending 1 task status or sending many task statuses at once. There's not really any added complexity here: you use the one that takes a map if you have more than 1 update, or you use the one that takes an id and a status if you have just 1. After that, all of the batching is handled for you, and it's handled in only 1 place, the `Reporter`.

anshulpundir commented:
I think keeping a single API which let’s you report multiple statuses would be fine. I think a map allocation for a single status should have negligible overhead (if golang does it well). 

I think this would address the concern about propagating the API change through since there would be only one API. 

It is not obvious to me what the motivation is behind keeping status reporter to only report a single status at a time. 

So, I’d rather push for a single API which let’s your report 1 or more statuses. If not, I'd suggest keeping the current change and reduce it down to a single which can report one or more statuses in a later PR.

stevvooe commented on agent/reporter.go:
```diff
@@ -76,6 +117,7 @@ func (sr *statusReporter) Close() error {
 }
 
 func (sr *statusReporter) run(ctx context.Context) {
+	ctx = log.WithModule(ctx, "statusReporter")
```
No camelCase in module names. You can just call this `reporter` or `status`.

stevvooe commented:
> It is not obvious to me what the motivation is behind keeping status reporter to only report a single status at a time.

This is about simplicity and granularity of error handling. If one only needs to worry about whether a single status failed or not, the retry algorithm can be as granular as need be. If it is batched together, retry needs to be built in for whole batches.

Looking at this change in detail, the unary `UpdateTaskStatus` is indeed the correct approach, as it allows the batching to be deferred to point where it can be most sensibly optimized.

dperny commented:
@stevvooe @anshulpundir @cyli changed to 1 method, PTAL.

anshulpundir commented on agent/reporter.go:
```diff
@@ -12,13 +12,21 @@ import (
 // concurrently, so implementations should be goroutine-safe.
 type StatusReporter interface {
-	UpdateTaskStatus(ctx context.Context, taskID string, status *api.TaskStatus) error
+	UpdateTaskStatuses(ctx context.Context, statuses map[string]*api.TaskStatus) error
```
this is bikeshedding, but rename to UpdateTaskStatus ?

stevvooe commented on agent/reporter.go:
```diff
@@ -102,27 +124,30 @@ func (sr *statusReporter) run(ctx context.Context) {
+		// swap out the statuses map, so we can release the lock while we
+		// do the batch
+		statuses := sr.statuses
+		sr.statuses = map[string]*api.TaskStatus{}
```
Because this copies the old status, an out of date copy has the potential to come through.

dperny replied:
sure, but that was the case before this change as well. if a new status comes in, then it will get addressed in the next go-round.

stevvooe commented on agent/reporter.go:
```diff
@@ -102,27 +124,30 @@ func (sr *statusReporter) run(ctx context.Context) {
+		// unimportant. we might add some statuses back to the map, which would
+		// be useless, but when we back get around to the sr.closed check
+		// above, we'll exit anyway.
+		if err != nil {
```
The granularity is going to problematic here: there are lots of cases where a single task status could result in an error, causing the group to get retried forever. For example, "not found" for a task status update should result in dropping the status, but that would result in a retry of the group.

dperny replied:
i recognize this. there's a problem of granularity in the opposite direction too, though, where if multiple statuses are bad, we keep having to retry.

dperny replied:
i'd like your rough take on solution.

dperny replied:
Oh, FWIW, the only error that can result from a bad task is if a node tries to send an update for a task it is not assigned. We can fossilize this in the API if we'd like.

dperny commented:
closing in favor of #2533
